### PR TITLE
Staking from smart contracts

### PR DESCRIPTION
**Pull request by weiribao**
This pull request allows staking Theta/TFuel from smart contracts.

### CODE REVIEW

File: ledger/vm/vm.go [hunk 1]
```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
...
+	expectedStake := big.NewInt(0).Add(currentStake, amount)
+	if expectedStake.Cmp(maxEliteEdgeNodeStake) > 0 {
+		return false
+	}
+
+	if db.GetBalance(sender).Cmp(amount) < 0 {
+		return false
+	}
+
+	if !checkBlsSummary(blsPubkey, blsPop, holderSig, eenAddr) {
```
Comment (jieyilong) [inline ledger/vm/vm.go:209]: In tx_deposit_stake.go seems we only check the BLS summary when the current EENP doesn't contain the holder address. Here we want to be more conservative?

cf.: https://github.com/thetatoken/theta-protocol-ledger/blob/583dea58dffe61691a593238145038a63bdc738e/ledger/execution/tx_deposit_stake.go#L196

File: ledger/vm/vm.go [hunk 2]
```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
...
+	view.UpdateGuardianCandidatePool(gcp)
+
+	return true
+}
+
+// UnstakeFromGuardian unstake from Guardians.
+func UnstakeFromGuardian(db StateDB, addr common.Address, guardianAddr common.Address) bool {
+	view := db.(*state.StoreView)
+	gcp := view.GetGuardianCandidatePool()
+	currentHeight := view.GetBlockHeight()
```
Comment (jieyilong) [inline ledger/vm/vm.go:166]: tx_withdraw_stake.go was using exec.state.Height() as the current height (which is actually off-by-one). Do we want to keep the behavior consistent (although view.GetBlockHeight() returns the correct height)?

https://github.com/thetatoken/theta-protocol-ledger/blob/583dea58dffe61691a593238145038a63bdc738e/ledger/execution/tx_withdraw_stake.go#L100

tx_deposit_stake.go was using the correct heights though.

File: ledger/vm/vm.go [hunk 3]
```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
...
+	return true
+}
+
+// UnstakeFromEEN unstake from EEN.
+func UnstakeFromEEN(db StateDB, addr common.Address, eenAddr common.Address) bool {
+	view := db.(*state.StoreView)
+
+	eenp := state.NewEliteEdgeNodePool(view, false)
+
+	currentHeight := view.GetBlockHeight()
```
Comment (jieyilong) [inline ledger/vm/vm.go:227]: Same as L166, exec.state.Height() vs view.GetBlockHeight()

File: ledger/vm/vm.go (hunk 2 above)
Comment (weiribao) [reply ledger/vm/vm.go]: updated

File: ledger/vm/vm.go (hunk 1 above)
Comment (weiribao) [reply ledger/vm/vm.go]: updated

File: ledger/vm/vm.go (hunk 3 above)
Comment (weiribao) [reply ledger/vm/vm.go]: updated.

File: ledger/vm/vm.go [hunk 4]
```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
...
+
+	db.SubThetaBalance(sender, amount)
+
+	err := gcp.DepositStake(sender, guardianAddr, amount, blsPubkey, view.GetBlockHeight())
+	if err != nil {
+		return false
+	}
+
+	view.UpdateGuardianCandidatePool(gcp)
+
```
[the remark this answers is not in the document]
Comment (jieyilong) [reply ledger/vm/vm.go:158]: Oh I was thinking the contract that calls the precompiled, but that probably has been handled by the EVM.

**Pull request merged**